libs/utils/src/tracing_span_assert.rs

```diff
@@ -71,58 +70,102 @@ impl<const L: usize> Extractor for MultiNameExtractor<L> {
-        self.as_ptr() == other.as_ptr()
-    }
-}
-impl<'a> Eq for MemoryIdentity<'a> {}
-impl<'a> Hash for MemoryIdentity<'a> {
-    fn hash<H: Hasher>(&self, state: &mut H) {
-        self.as_ptr().hash(state);
-    }
-}
-impl<'a> fmt::Debug for MemoryIdentity<'a> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{:p}: {}", self.as_ptr(), self.0.name())
-    }
-}
-
-/// The extractor names passed as keys to [`new`].
+/// Checks that the given extractors are satisfied with the current span hierarchy.
+///
+/// `Ok` is also returned when tracing is not configured.
 pub fn check_fields_present<const L: usize>(
     must_be_present: [&dyn Extractor; L],
-) -> Result<(), Vec<&dyn Extractor>> {
-    let mut missing: HashSet<MemoryIdentity> =
-        HashSet::from_iter(must_be_present.into_iter().map(|r| MemoryIdentity(r)));
+) -> Result<Summary, Vec<&dyn Extractor>> {
+    let mut missing = must_be_present.into_iter().collect::<Vec<_>>();
     let trace = tracing_error::SpanTrace::capture();
     trace.with_spans(|md, _formatted_fields| {
-        missing.retain(|extractor| match extractor.0.extract(md.fields()) {
+        missing.retain(|extractor| match extractor.extract(md.fields()) {
             ExtractionResult::Present => false,
             ExtractionResult::Absent => true,
         });
         !missing.is_empty() // continue walking up until we've found all missing
     });
     if missing.is_empty() {
-        Ok(())
+        Ok(Summary::FoundEverything)
+    } else if !tracing_subscriber_configured() {
+        Ok(Summary::Unconfigured)
     } else {
-        Err(missing.into_iter().map(|mi| mi.0).collect())
+        // we can still hit here if a tracing subscriber has been configured but the ErrorLayer is
+        // missing, which can be annoying. for this case, we could probably use
+        // SpanTrace::status().
+        Err(missing)
     }
 }
 
+/// Explanation for why the check was deemed ok.
+///
+/// Mainly useful for testing.
+#[derive(Debug)]
+pub enum Summary {
+    /// All extractors were found.
+    ///
+    /// Should only happen when tracing is properly configured.
+    FoundEverything,
+
+    /// Tracing has not been configured at all. This is ok for tests running without tracing set
+    /// up.
+    Unconfigured,
+}
+
+fn tracing_subscriber_configured() -> bool {
+    // SpanTrace::status() is not a strong indicator because it short circuits to EMPTY on empty
+    // spans, so let's check if there's an ErrorLayer directly. alas, we cannot due to
+    // pub(crate) but we can check if the subscriber is the noop impl.
+
+    let span = tracing::Span::current();
+    let mut configured_at_all = false;
+    span.with_subscriber(|(_, s)| {
+        // assume that if there is a non-NoSubscriber, it's a proper one
+        configured_at_all = s
+            .downcast_ref::<tracing::subscriber::NoSubscriber>()
+            .is_none();
+    });
+
+    configured_at_all
+}
```

Comment:
This is wrong; now I understand why SpanTrace::status defaults to empty, because that is the only deduction we can do, as the dispatch linked to the span will be nothing if the span has been filtered out.